File: src/main/scala/algorithms/searching/BinarySearch.scala
package algorithms.searching

object BinarySearch {

  def main(args: Array[String]): Unit = {
    val arr1 = Array(1,2,3,4,5,6,7)
    println(bSearch(3, arr1))

    val arr2 = Array(55, 95, 121, 340)
    println(bSearch(242, arr2))

    val arr3 = Array(55, 95, 121, 340)
    println(bSearch(41, arr3))

    val arr4 = Array(55, 95, 121, 340)
    println(bSearch(441, arr4))
  }

  def bSearch(value: Int, arr: Array[Int]): Int = {

    var start = 0
    var end = arr.size - 1
    var mid = start + ((end-start)/2)

    while (end > start){

      val midVal = arr(mid)

      if (arr(end) == value)
        return end
      else if (arr(start) == value)
        return start
      else if (arr(mid) == value)
        return mid
      else{
        if ((end - start)<2)
          return -1
        if (value < midVal)
          end = mid
        else
          start = mid
        mid = start + ((end-start)/2)
      }
    }

    return -1
  }
}

File: src/main/python/algorithms/searching/binary_search.py
"""
Given a sorted array, search the position of an element
"""


def binary_search(arr, num_to_search, start):

    if num_to_search > arr[len(arr)-1] or num_to_search < arr[0]:
        return -1
    mid = int(len(arr)/2)

    if num_to_search == arr[mid]:
        return start + mid
    if num_to_search < arr[mid]:
        return binary_search(arr[:mid], num_to_search, start)
    else:
        return binary_search(arr[mid:], num_to_search, start+mid)


if __name__ == "__main__":

    print(binary_search([1, 2, 3, 4, 5], 6, 0))
    print(binary_search([1, 2, 3, 4, 5], 0, 0))
    print(binary_search([1, 2, 3, 4, 5], 1, 0))
    print(binary_search([1, 2, 3, 4, 5], 3, 0))
    print(binary_search([1, 2, 3, 4, 5], 4, 0))
    print(binary_search([1, 2, 3, 4, 5], 5, 0))
    print(binary_search([1, 2, 3, 4], 1, 0))
    print(binary_search([1, 2, 3, 4], 2, 0))
    print(binary_search([1, 2, 3, 4], 3, 0))
    print(binary_search([1, 2, 3, 4], 4, 0))
